src/hotspot/share/opto/addnode.cpp

```diff
@@ -1004,16 +991,29 @@ const Type *XorINode::add_ring( const Type *t0, const Type *t1 ) const {
   const TypeInt *r0 = t0->is_int(); // Handy access
   const TypeInt *r1 = t1->is_int();
 
-  // Complementing a boolean?
-  if( r0 == TypeInt::BOOL && ( r1 == TypeInt::ONE
-                               || r1 == TypeInt::BOOL))
-    return TypeInt::BOOL;
+  if (r0->is_con() && r1->is_con()) {
+    // just XOR them bits.
+    return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  }
+
+  // not constants
 
-  if( !r0->is_con() || !r1->is_con() ) // Not constants
-    return TypeInt::INT;        // Any integer, but still no symbols.
+  // result of xor can only have bits sets where any of the
+  // inputs have bits set. lo can always become 0.
 
-  // Otherwise just XOR them bits.
-  return TypeInt::make( r0->get_con() ^ r1->get_con() );
+  if (r0->_lo >= 0 && r1->_lo >= 0) {
+      // x ^ y cannot have any bit set that is higher than both the highest bits set in x and y
+      // x cannot have any bit set that is higher than the highest bit set in r0->_hi
+      // y cannot have any bit set that is higher than the highest bit set in r1->_hi
+
+      // note cast to unsigned happens before +1 to avoid signed overflow, and
+      // round_up is safe because high bit is unset (0 <= lo <= hi)
```

Comment:
```suggestion
      // Note: cast to unsigned happens before +1 to avoid signed overflow, and
      // round_up is safe because high bit is unset (0 <= lo <= hi)
```